modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/NettyTcpChannel.java

```diff
@@ -91,6 +95,11 @@ public void sendMessage(BytesReference reference, ActionListener<Void> listener)
                 listener.onFailure((Exception) cause);
             }
         });
+        channel.writeAndFlush(Netty4Utils.toByteBuf(reference), writePromise);
+        
+        if (channel.eventLoop().isShutdown()) {
```

Comment:
do we protect this from double invocation? I think we should just make sure we only invoke once. can you wrap it in here and ensure that? maybe just use `NotifyOnceListener`

Reply:
Well - the only place we call this is from `internalSendMessage`. 

```
    private void internalSendMessage(TcpChannel channel, BytesReference message, SendMetricListener listener) {
        try {
            channel.sendMessage(message, listener);
        } catch (Exception ex) {
            // call listener to ensure that any resources are released
            listener.onFailure(ex);
            onException(channel, ex);
        }
    }
```

And `SendMetricListener` is a `NotifyOnceListener`. So incidentally - yes it is always a notify once listener. Do you want me to codify that by changing the signature of `TcpChannel#sendMessage` to:

```
/**
     * Sends a tcp message to the channel. The listener will be executed once the send process has been
     * completed.
     *
     * @param reference to send to channel
     * @param listener to execute upon send completion
     */
    void sendMessage(BytesReference reference, NotifyOnceListener<Void> listener);
```

???

Reply:
@s1monw

Reply:
it's find in this case! LGTM